src/test/scala/pt/tecnico/dsi/neutron/CrudSpec.scala

```diff
@@ -1,65 +1,69 @@
 package pt.tecnico.dsi.neutron
 
-import cats.effect.IO
+import cats.effect.{IO, Resource}
 import cats.implicits._
-import cats.syntax.flatMap._
 import io.circe.{Decoder, Encoder}
-import pt.tecnico.dsi.neutron.models.Model
+import org.scalatest.Assertion
+import pt.tecnico.dsi.neutron.models.{Model, Network}
 import pt.tecnico.dsi.neutron.services.{BulkCreate, CrudService}
 import pt.tecnico.dsi.openstack.common.models.WithId
 
-abstract class CrudSpec[T <: Model](val name: String, val service: NeutronClient[IO] => CrudService[IO, T])
+abstract class CrudSpec[T <: Model](val name: String)
   (implicit val encoder: Encoder[T#Create], val decoder: Decoder[WithId[T#Read]])
   extends Utils {
 
+  val service: CrudService[IO, T]
+
   val displayName: String = name.capitalize
-  val createStub: IO[T#Create]
+  val updateStub: IO[T#Update]
+  def updateComparator(read: T#Read, update: T#Update): Assertion
 
-  val withSubCreated: IO[(WithId[T#Read], CrudService[IO, T])] =
-    for {
-      neutron <- client
-      stub <- createStub
-      crudService = service(neutron)
-      createdStub <- crudService.create(stub)
-    } yield (createdStub, crudService)
+  val withStubCreated: Resource[IO, WithId[T#Read]]
 
+  val withNetworkCreated: Resource[IO, WithId[Network.Read]] = {
+    val created = client.networks.create { Network.Create() }
+    Resource.make(created)(x => client.networks.delete(x.id))
+  }
 
   s"$displayName service" should {
 
-    "create and get" in withSubCreated .flatMap { case (stub, service) =>
-      service.get(stub.id) .idempotently(_ shouldBe stub)
+    "create and get" in withStubCreated.use[IO, Assertion] {
+      stub => service.get(stub.id).idempotently(_ shouldBe stub)
     }
 
-    "delete" in {
+    "delete" in withStubCreated.use[IO, Assertion] { stub =>
       for {
-        (stub, service) <- withSubCreated
         _ <- service.delete(stub.id)
         lst <- service.list().compile.toList
       } yield assert(!lst.exists(_.id == stub.id))
     }
 
-    "list" in {
+    "list" in withStubCreated.use[IO, Assertion] { stub =>
       for {
-        (stub, service) <- withSubCreated
         lst <- service.list().compile.toList
       } yield assert(lst.exists(_.id == stub.id))
     }
+
+    "update" in withStubCreated.use[IO, Assertion] { stub =>
+      for {
+        ustub <- updateStub
+        updated <- service.update(stub.id, ustub)
+      } yield updateComparator(updated, ustub)
+    }
   }
 }
 
 trait BulkCreateSpec[T <: Model] {
   self: CrudSpec[T] =>
 
-  val bulkService: NeutronClient[IO] => BulkCreate[IO, T]
+  val service: CrudService[IO, T] with BulkCreate[IO, T]
+  def withBulkCreated(quantity: Int): Resource[IO, List[WithId[T#Read]]]
```

Comment:
`def withBulkCreated(quantity: Int = 5): Resource[IO, List[WithId[T#Read]]]`
and remove:
`val n = 5`